### PR TITLE
refactor profile validation code

### LINKED ISSUE #1951: refactor profile validation code
We could do a lot better with code reuse on the profile validation code.

### PR DESCRIPTION
**Pull request by aliceriot**
#### What are the relevant tickets?

closes #1951 

#### What's this PR do?

This rewrites the profile validation code in a more functional style, which lets us have a greater degree of code reuse and, hopefully, readability.

#### Where should the reviewer start?

Read over the changes, and make sure they make sense.

#### How should this be manually tested?

Well, I think just make sure there aren't any regressions in the validation UX, when filling out the profile.

### CODE REVIEW

**Comment by noisecapella** [comment]
Whoops

File: static/js/lib/validation/profile_test.js [hunk 1]
```diff
@@ -166,13 +166,13 @@ describe('Profile validation functions', () => {
...
           company_name: 'Name of Employer is required'
-        }]
+        },{}]
       }, employmentValidation(clone));
     });
 
     it('should reject end date before start date', () => {
       let errors = {
-        work_history: [, // eslint-disable-line no-sparse-arrays
+        work_history: [{}, // eslint-disable-line no-sparse-arrays
```
Comment (noisecapella) [inline static/js/lib/validation/profile_test.js]: Can you remove the `eslint-disable` here and in the next changes?

Comment (aliceriot) [inline static/js/lib/validation/profile_test.js]: oh yeah, oops. no longer necessary!

File: static/js/lib/validation/profile.js [hunk 2]
```diff
@@ -57,148 +47,164 @@ export function programValidation(_: Profile, ui: UIState): ValidationErrors {
...
+  'city': 'City is required',
+  'country': 'Country is required',
+  'state_or_territory': 'State or Territory is required',
+};
+
+// functions to perform extra checks
+// must be binary functions taking (entry, errors)
+// where entry is a work history entry
+// and errors is the output of workHistoryErrors
+// and returning errors U newErrors
```
Comment (noisecapella) [inline static/js/lib/validation/profile.js]: Pretty trivial but can you change `U` to `∪`?

Comment (aliceriot) [inline static/js/lib/validation/profile.js]: Sure, I meant the set union character here, so as long as that's clear.

Comment (aliceriot) [inline static/js/lib/validation/profile.js]: (well, sort of the set union, objects aren't really sets at all, haha)

Comment (noisecapella) [inline static/js/lib/validation/profile.js]: Yeah, I took that character from the set union wiki page

Comment (aliceriot) [inline static/js/lib/validation/profile.js]: oh nifty! and here I thought it was just a `u`

**Comment by noisecapella** [comment]
:+1:

**Pull request merged**